sec1_intro_and_setup/test_app/.gitattributes

```diff
@@ -0,0 +1,10 @@
+# See https://git-scm.com/docs/gitattributes for more about git attribute files.
```

Comment:
The `.gitattributes` file allows one to specify the files and paths attributes that should be used by git when performing git actions, such as `git commit`, etc. It automatically saves the file according to the attributes specified every time a file is created or saved. Not all developers use the same dev environment (for example, some use Window machine VScode, some use MAC  machine Sublime Text), so some of the attributes are different (for example, end of line, Windows' eol is CRLF while MAC's eol is LF). The attributes difference could cause errors. `.gitattributes` ensures every developer uses the same value when using different machines / OSes across the repository.

Reply:
can you give an example of how this is used?

Reply:
you mean the command to create the .gitattribute file?

Reply:
I'm not sure if there is a command for this.
ok. Let me rephrase.

Can you speak in english what this file is about?

Reply:
This `.gitattributes` file is about giving attributes to pathnames. The `.gitattributes` file is needed to ensure every developer uses the same value when using different machines / OSes across the repository. To check whether an attribute is given to a pathname

Reply:
Ok. Then, in the 3 that is stated in this file, what attributes is given? and why give them here?

Reply:
```

Reply:
You still didn't answer my question on why give them here.

Reply:
Why these 3 are given attributes to pathnames?:

When more than one pattern matches the path, a later line overrides an earlier line. This overriding is done per attribute. Attributes that should be version-controlled and distributed to other repositories (i.e., attributes of interest to all users) should go into .gitattributes files. For example, 

- `db/schema.rb` is used to capture the current state of the database schema after a migration. The problem is there is a high possibility of merging conflict when there are 2 different branches modifying schema in the development tree, so the `db/schema.rb` is put in`.gitattribute` file to keep the `db/schema.rb` in order and combining independent migrations without errors induced by merging, 

so that developers with different machine / OSes can use the same value accross the repository.

---------------------------------
Question:
1. Why most people put `db/schema.rb` in `.gitignore` instead of `.gitattributes`?

Reply:
> db/schema.rb is used to capture the current state of the database schema after a migration. The problem is there is a high possibility of merging conflict when there are 2 different branches modifying schema in the development tree, so the db/schema.rb is put in.gitattribute file to keep the db/schema.rb in order and combining independent migrations without errors induced by merging,

This is incorrect. To answer this, you need to know about `db/schema.rb`

---

Your Question:
1. Who suggest to do that? The ones who do that are the ones who have no idea what they are doing

Reply:
Why give `linguist-generated` attribute to `db/schema.rb` and `yarn.lock`?
- `db/schema.rb` - The `schema.rb` file documents the final current state of the database schema and is used by the `tests` to populate the database schema. The `schema.rb` file is used for migrations which allows one to alter database in a structured and organized manner. However, the `schema.rb` file often creates large diffs and merge conflicts as it is updated. `linguist-generated` allows one to suppress the display and merge conflicts of `db/schema.rb` in code review by using `git diff`.

- `yarn.lock` is large. `linguist-generated` is significant it comes to suppressing the display in code review.

Why give `linguist-vendored` attribute to `vendor/.keep`?
- Checking code you didn't write into your git repo is a common practice, but this often inflates the project's language stats and may even cause your project to be labelled as another language. By default, Linguist treats all of the paths defined in `vendor.yml` as vendored and therefore does not include them in the language statistics for a repository. Hence, `linguist-vendored` attribute is used to vendor or un-vendor paths.

Reply:
- For `db/schema.rb`, if you are going to be adding this to git, why want to surpressed the merge conflicts here?
- ok

Reply:
- Suppressing the merge conflicts makes it easier to read, so that one does not need to go through the whole file if the commit is large.

Reply:
How does it make it easier to read?

Reply:
- If one is using GitHub and work with generated files in the repository, one may be seeing large diffs and merge conflicts as one is updating newly generated files. . Using `.gitattributes` file in the repository can suppress the display and merge conflicts of generated files in code review by putting `linguist-generated=true` in the `.gitattributes` file. The `-diff` option tells git to treat the file like a binary file. The `diff` output will indicate that the file has changed but won't display the differences. The `-merge` option tells Git to treat merging like a binary. It will accept the latest version and avoid any conflict resolution. The command line git lets you know that the files have changed but does not show the contents. Also the merge option will prevent git from prompting you about merge conflicts which can happen with generated files.

Reply:
i see.
But, lets take the example of db/schema.rb

Lets say 2 devs:
dev1 update db/schema.rb with stuffs
dev2 update db/schema.rb with stuffs

dev1 merge his code into main
dev2 pull dev1 changes into dev2 branch. Without this setting, it will be merge conflict. With this settings, it will take the latest one which is dev2. Dev2 db/schema.rb do not have changes by dev1. Then at the end, it will cause problem because it only container dev2 codes but not dev1. 

How is it better?

Reply:
Can you explain in details about this:
>dev2 pull dev1 changes into dev2 branch. Without this setting, it will be merge conflict. With this settings, it will take the latest one which is dev2. Dev2 db/schema.rb do not have changes by dev1. Then at the end, it will cause problem because it only container dev2 codes but not dev1.

Reply:
What is it you want me to explain about?

Reply:
You mention `dev1 merge his code into main` then `dev2 pull dev1 changes into dev2 branch`, does this mean, dev2 pull the merged main or dev2 pull dev1 before dev1 merge into main?

Reply:
in specific git terms:

- we have main
- dev1 have branch1 (branch from main)
- dev2 have branch2 (branch from main)

Flow:
1. dev1 merge branch1 into main using PR
2. Now main have branch1 code due to (1)
3. dev2 wants branch1 code which is in main into branch2
4. dev2 checkout to dev2 local main
5. dev2 (in local main) pull down changes in github main
6. dev2 checkout to local branch2
7. dev2 merge local main to local branch2

This is what i mean.

Reply:
So the `without this setting`, the `this` refers to?

Reply:
```
without this setting
```
referring to adding the db/schema.rb to this file

Reply:
Let me get the idea of the question first:
> dev1 merge his code into main
dev2 pull dev1 changes into dev2 branch. Without this setting, it will be merge conflict. With this settings, it will take the latest one which is dev2. Dev2 db/schema.rb do not have changes by dev1. Then at the end, it will cause problem because it only container dev2 codes but not dev1.

The question means:
1. dev2 pull from merged main (main + dev1 changes)
2. `Without this setting, it will be merge conflict` means the merged contains main and dev1 changes but without dev1's `db/schema.rb`, and this will prompt merge conflict.
3. `With this settings, it will take the latest one which is dev2` means `main + dev1 changes + dev1's db/schema.rb`, the latest version of the repo is dev2.
4. I don't understand this: `Dev2 db/schema.rb do not have changes by dev1. Then at the end, it will cause problem because it only container dev2 codes but not dev1.` What does this mean? Right now, you are `with adding dev1's db/schema.rb` or `without adding dev1's db/schema.rb` before merging dev1 to main?

Reply:
```
Dev2 db/schema.rb do not have changes by dev1. Then at the end, it will cause problem because it only container dev2 codes but not dev1.
```

This happens when the merge happens and dev2 code is taken for the schema.rb without dev1 codes. So, although dev2 merged dev1 changes, it won't have dev1 codes in schema.rb

Reply:
Suppressing the changes here does not make the scenario better since dev2 does not have dev1's schema.rb.

Question:
1. If based on the scenario given, it is better for dev2 to clone the merged repo and do dev2's code all over again?

Reply:
1. No. Imagine you are working on a feature. Then the issue happen. Then you are told to remove your code then reclone then redo your code again. It will be frustrating isn't it? Now, multiply that over 60 people who is working on the same codebase. Will that even be viable to do?

Reply:
No

Reply:
So, then how?

Reply:
The steps:
1. Make a backup of your current code
2. Follow the git suggestions to fix the git conflict, then merge
3. Pull the merged repo 
4. Replace the changes that arise merge conflict before merging (in 2.) (should be no merge conflict this time)

Reply:
Based on what you said, if you put db/schema here, it won't show merge conflict isn't it?

Reply:
Yes

Reply:
???

Reply:
What's the question?

Reply:
Please read back up the thread

Reply:
> Based on what you said, if you put db/schema here, it won't show merge conflict isn't it?

If merge conflict appears in db/shcema.rb, use `merge_schema_rb` to resolve the merge conflicts.

Reply:
????

Focus on the main question. You said don't put it here. I said put it here. Thats the entire argument. What to do with if happen merge conflict or not?

Reply:
Do migration for `schema.rb`.

> What is migration?
Migrations allow you to use Ruby programs to modify the database structure. Migrations help you to track the modifications implemented and leave a record calls `migration record`. You can add a migration file with `rake db:migrate` and it will figure out which migrations to run by itself so that all developers and officially deployed servers can easily synchronize the latest database structure.

Steps when merge conflict happens:
1. Pull the latest code in the main branch.
2. Run `rails db:migrate:status` to check the migration status of the main branch. (`up` describe the transformation you’d like to make to your schema, and the `down` method of your migration should revert the transformations done by the up method. )
3. Switch to the branch you are working on (we call it `branch1` here for example)
4. Try `git merge main` but got merge conflict (from here, we know whether the merge conflict happens in the schema.rb. If the merge conflict happens in schema.rb, perform the following steps)
5. We check the migration status of branch1 schema.rb and compare (we will see where the merge conflicts here, the `up` and `down`)
6. We reset to the schema.rb version before merge conflict happens (back to the `up` before `down` in the migration status) by `git reset HEAD db/schema.rb`, then we unstage the changes
6. We do the migration with `rails db:migrate`
7. Check the migration status of branch1 schema.rb with `rails db:migrate:status` before (merging it to the main branch/ stage the changes)

Reply:
Not 100% accurate but, lets move on. You'll understand better later